src/Cache/Context/OgMembershipStateCacheContext.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+namespace Drupal\og\Cache\Context;
+
+use Drupal\Core\Cache\CacheableMetadata;
+use Drupal\Core\Cache\Context\CacheContextInterface;
+use Drupal\Core\Routing\RouteMatchInterface;
+use Drupal\Core\Session\AccountInterface;
+use Drupal\og\GroupTypeManager;
+use Drupal\og\MembershipManagerInterface;
+use Drupal\og\OgMembershipInterface;
+
+/**
+ * Defines a cache context service, for "membership state" caching.
+ *
+ * Cache context ID: 'og_membership_state'
+ */
+class OgMembershipStateCacheContext implements CacheContextInterface {
+
+
+  /**
+   * The group type manager service.
+   *
+   * @var \Drupal\og\GroupTypeManager
+   */
+  protected $groupTypeManager;
+
+
+  /**
+   * The membership manager service.
+   *
+   * @var \Drupal\og\MembershipManagerInterface
+   */
+  protected $membershipManager;
+
+  /**
+   * The route match service.
+   *
+   * @var \Drupal\Core\Routing\RouteMatchInterface
+   */
+  protected $routeMatch;
+
+  /**
+   * The current user.
+   *
+   * @var \Drupal\Core\Session\AccountInterface
+   */
+  protected $user;
+
+  /**
+   * Constructs a new UserCacheContextBase class.
+   *
+   * @param \Drupal\Core\Session\AccountInterface $user
+   *   The current user.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The current route match service.
+   * @param \Drupal\og\GroupTypeManager $group_type_manager
+   *   The group type manager service.
+   * @param \Drupal\og\MembershipManagerInterface $membership_manager
+   *   The membership manager service.
+   */
+  public function __construct(AccountInterface $user, RouteMatchInterface $route_match, GroupTypeManager $group_type_manager, MembershipManagerInterface $membership_manager) {
+    $this->user = $user;
+    $this->routeMatch = $route_match;
+    $this->groupTypeManager = $group_type_manager;
+    $this->membershipManager = $membership_manager;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function getLabel() {
+    return t('OG membership state');
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getContext() {
+    if (!$route_contexts = $this->routeMatch->getRouteObject()->getOption('parameters')) {
+      // No "parameters" defined in the route.
+      return 'none';
+    }
```

Comment:
Yeah, OG context might be needed after all. /cc @RoySegall

On Sep 7, 2016 3:25 PM, "Pieter Frenssen" <notifications@github.com> wrote:

> In src/Cache/Context/OgMembershipStateCacheContext.php
> <https://github.com/Gizra/og/pull/163#discussion_r77811136>:
>
> > +
> > +  /**
> > +   * {@inheritdoc}
> > +   */
> > +  public static function getLabel() {
> > +    return t('OG membership state');
> > +  }
> > +
> > +  /**
> > +   * {@inheritdoc}
> > +   */
> > +  public function getContext() {
> > +    if (!$route_contexts = $this->routeMatch->getRouteObject()->getOption('parameters')) {
> > +      // No "parameters" defined in the route.
> > +      return 'none';
> > +    }
>
> Well now it only works on a very specific subset of routes that follow
> very specific rules (like being formatted as node/%). By leveraging the
> OG context it would work out of the box on all routes that have a group
> associated with them. That sounds more useful to me. It would of course
> also work on node/%).
>
> I will have to look into this though, I start to understand caching a bit
> better lately but I am still unsure about some aspects of it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Gizra/og/pull/163/files/ad169ad87998684209857708e306f07c7e3f4f14#r77811136>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAHrC0GOkqYzFCV9gjj8r2ZmHigaKB-Gks5qnq0jgaJpZM4J0nZ->
> .
>